src/miner.cpp

```diff
@@ -707,6 +727,48 @@ CBlockTemplate* CreateNewBlock(const CScript& scriptPubKeyIn,  unsigned int nBlo
                 continue;
             }
 
+            // Skip transaction if we cannot add it to the sc commitment tree
+            if (pblock->nVersion == BLOCK_VERSION_SC_SUPPORT) {
+                // Check tx commitment tree limits
+                bool scCommitGuardRes;
+                if (tx.IsCertificate()) {
+                    scCommitGuardRes = scCommGuard.add(dynamic_cast<const CScCertificate&>(tx));
+                } else {
+                    // Immediately restore scCommitmentGuard to a valid state in case of failure
+                    scCommitGuardRes = scCommGuard.add(dynamic_cast<const CTransaction&>(tx), true);
+                }
+
+                if (!scCommitGuardRes) {
+                    LogPrint("sc", "%s():%d - Skipping [%s] because txs commitment tree guard failed\n",
+                            __func__, __LINE__, tx.GetHash().ToString());
+                    continue;
+                }
+
+                // Try adding to the real commitment tree if previous step was successful
+                bool scCommitmentBuilderResult;
+                if (tx.IsCertificate()) {
+                    scCommitmentBuilderResult = scCommBuilder->add(dynamic_cast<const CScCertificate&>(tx), view);
+                }
+                else {
+                    scCommitmentBuilderResult = scCommBuilder->add(dynamic_cast<const CTransaction&>(tx));
+                }
+                if (!scCommitmentBuilderResult)
+                {
+                    LogPrint("sc", "%s():%d - Skipping [%s] because we cannot add that to the sc commitment tree\n",
```

Comment:
I'd say also `scCommGuard` has to rebuilt (or its counters be updated) in case there is a failure on `scCommBuilder`.

Reply:
Yes, you are right! I modified this section to keep them aligned.